File: app/src/main/java/com/donald/musictheoryapp/Music/ScoreView/SAccidental.java
package com.donald.musictheoryapp.Music.ScoreView;

import android.graphics.Canvas;

import com.donald.musictheoryapp.Music.MusicXML.Clef;
import com.donald.musictheoryapp.Music.MusicXML.Note;

public class SAccidental extends ScoreStamp
{
	private char glyph;
	private float glyphWidth;
	private float verticalOffsetOnStaff;
	//private final float horizontalOffsetOnStaff;

	SAccidental(ScoreView view)
	{
		super(view);
		glyph = U.ACC_NATURAL;
		verticalOffsetOnStaff = 0;
		//horizontalOffsetOnStaff = -(noteWidth() + noteWidth() * 0.5f + paddingFromNote);
	}

	protected void setNote(Note note, Clef clef)
	{
		verticalOffsetOnStaff = -Clef.noteStaffPosition(note, clef) * staffStepHeight() - staffHeight();
		switch(note.accidental)
		{
			case Note.Accidental.FLAT_FLAT:
				glyph = U.ACC_D_FLAT;
				glyphWidth = noteWidth() * 1.3f;
				break;
			case Note.Accidental.FLAT:
				glyph = U.ACC_FLAT;
				glyphWidth = noteWidth() * 0.8f;
				break;
			case Note.Accidental.NATURAL:
				glyph = U.ACC_NATURAL;
				glyphWidth = noteWidth() * 0.8f;
				break;
			case Note.Accidental.SHARP:
				glyph = U.ACC_SHARP;
				glyphWidth = noteWidth() * 0.8f;
				break;
			case Note.Accidental.SHARP_SHARP:
				glyph = U.ACC_D_SHARP;
				glyphWidth = noteWidth() * 0.8f;
				break;
			case Note.Accidental.NULL:
				return;
			default:
				throw new IllegalArgumentException(note.pitch.alter + " + is not a valid accidental.");
		}
	}

	@Override
	protected void onDraw(Canvas canvas, float posX, float posY)
	{
		float paddingFromNote = noteWidth() * 0.5f;

		canvas.drawText(String.valueOf(glyph),
			//posX + horizontalOffsetOnStaff,
			posX - glyphWidth - paddingFromNote - noteWidth() / 2,
			posY + glyphSize() + verticalOffsetOnStaff,
			glyphPaint()
		);
	}
}

File: app/src/main/java/com/donald/musictheoryapp/music/scoreview/SAccidental.kt
package com.donald.musictheoryapp.music.scoreview

import android.graphics.Canvas
import com.donald.musictheoryapp.music.MusicXML.Clef
import com.donald.musictheoryapp.music.MusicXML.Clef.Companion.noteStaffPosition
import com.donald.musictheoryapp.music.MusicXML.Note
import java.lang.IllegalArgumentException

internal class SAccidental(view: ScoreView) : ScoreStamp(view) {

    private var glyph = U.ACC_NATURAL
    private var accRelWidth = 0F
    private var noteStaffPosition = 0
    private var draws = false
    override val relContentWidth: Float
        get() = accRelWidth + if (draws) REL_MARGIN_AFTER_NOTE_ACC else 0F

    var note: Note? = null
    var clef: Clef? = null

    fun set(note: Note, clef: Clef) {
        when (note.accidental) {
            Note.Accidental.FLAT_FLAT -> {
                glyph = U.ACC_D_FLAT
                accRelWidth = REL_DOUBLE_ACC_WIDTH //noteWidth() * 1.3f
                draws = true
            }
            Note.Accidental.FLAT -> {
                glyph = U.ACC_FLAT
                accRelWidth = REL_SINGLE_ACC_WIDTH //noteWidth() * 0.8f
                draws = true
            }
            Note.Accidental.NATURAL -> {
                glyph = U.ACC_NATURAL
                accRelWidth = REL_SINGLE_ACC_WIDTH //noteWidth() * 0.8f
                draws = true
            }
            Note.Accidental.SHARP -> {
                glyph = U.ACC_SHARP
                accRelWidth = REL_SINGLE_ACC_WIDTH //noteWidth() * 0.8f
                draws = true
            }
            Note.Accidental.SHARP_SHARP -> {
                glyph = U.ACC_D_SHARP
                accRelWidth = REL_SINGLE_ACC_WIDTH//noteWidth() * 0.8f
                draws = true
            }
            Note.Accidental.NULL -> {
                glyph = U.ACC_NATURAL
                accRelWidth = 0F
                draws = false
            }
            else -> {
                glyph = U.ACC_NATURAL
                accRelWidth = 0F
                draws = false
                throw IllegalArgumentException(note.pitch.alter.toString() + " + is not a valid accidental.")
            }
        }
        noteStaffPosition = noteStaffPosition(note, clef)
    }

    override fun onDraw(canvas: Canvas, posX: Float, posY: Float) {
        //val note = note ?: return
        //val clef = clef ?: return
        //set(note, clef)
        if (draws) {
            val verticalOffsetOnStaff = -noteStaffPosition * staffStepHeight + staffHeight
            canvas.drawText(
                glyph.toString(),
                posX,
                posY + verticalOffsetOnStaff,
                glyphPaint
            )
        }
    }

}